src/wp-includes/class-wp-application-passwords.php

```diff
@@ -0,0 +1,400 @@
+	 * @type string
+	 */
+	const USERMETA_KEY_APPLICATION_PASSWORDS = '_application_passwords';
+
+	/**
+	 * The length of generated application passwords.
+	 *
+	 * @type integer
+	 */
+	const PW_LENGTH = 24;
+
+	/**
+	 * Add various hooks.
+	 *
+	 * @since ?.?.0
+	 *
+	 * @access public
+	 * @static
+	 */
+	public static function add_hooks() {
+		add_filter( 'authenticate', array( __CLASS__, 'authenticate' ), 10, 3 );
+		add_action( 'rest_api_init', array( __CLASS__, 'rest_api_init' ) );
+		add_filter( 'determine_current_user', array( __CLASS__, 'rest_api_auth_handler' ), 20 );
+		self::fallback_populate_username_password();
+	}
+
+	/**
+	 * Handle declaration of REST API endpoints.
+	 *
+	 * @since ?.?.0
+	 *
+	 * @access public
+	 * @static
+	 */
+	public static function rest_api_init() {
+		// Some hosts that run PHP in FastCGI mode won't be given the Authentication header.
+		register_rest_route(
+			'2fa/v1',
+			'/test-basic-authorization-header/',
+			array(
+				'methods'             => WP_REST_Server::READABLE . ', ' . WP_REST_Server::CREATABLE,
+				'callback'            => __CLASS__ . '::rest_test_basic_authorization_header',
+				'permission_callback' => '__return_true',
+			)
+		);
+	}
+
+	/**
+	 * Loosely Based on https://github.com/WP-API/Basic-Auth/blob/master/basic-auth.php
+	 *
+	 * @since ?.?.0
+	 *
+	 * @access public
+	 * @static
+	 *
+	 * @param $input_user
+	 *
+	 * @return WP_User|bool
+	 */
+	public static function rest_api_auth_handler( $input_user ) {
+		// Don't authenticate twice.
+		if ( ! empty( $input_user ) ) {
+			return $input_user;
+		}
+
+		// Check that we're trying to authenticate
+		if ( ! isset( $_SERVER['PHP_AUTH_USER'] ) ) {
+			return $input_user;
+		}
+
+		$user = self::authenticate( $input_user, $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW'] );
+
+		if ( $user instanceof WP_User ) {
+			return $user->ID;
+		}
+
+		// If it wasn't a user what got returned, just pass on what we had received originally.
+		return $input_user;
+	}
+
+	/**
+	 * Test whether PHP can see Basic Authorization headers passed to the web server.
+	 *
+	 * @return WP_Error|array
+	 */
+	public static function rest_test_basic_authorization_header() {
+		$response = array();
+
+		if ( isset( $_SERVER['PHP_AUTH_USER'] ) ) {
+			$response['PHP_AUTH_USER'] = $_SERVER['PHP_AUTH_USER'];
+		}
+
+		if ( isset( $_SERVER['PHP_AUTH_PW'] ) ) {
+			$response['PHP_AUTH_PW'] = $_SERVER['PHP_AUTH_PW'];
+		}
+
+		if ( empty( $response ) ) {
+			return new WP_Error( 'no-credentials', __( 'No HTTP Basic Authorization credentials were found submitted with this request.' ), array( 'status' => 404 ) );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Some servers running in CGI or FastCGI mode don't pass the Authorization
+	 * header on to WordPress.  If it's been rewritten to the `REMOTE_USER` header,
+	 * fill in the proper $_SERVER variables instead.
+	 */
+	public static function fallback_populate_username_password() {
```

Comment:
We can move this to `wp_fix_server_vars`.

Reply:
Moved to `wp_populate_basic_auth_from_authorization_header()` that is called by `wp_fix_server_vars()`.